[JEP-229] Updates (#332)
jenkinsci/jep: jep/229/README.adoc

with a link:https://github.com/jglick/jenkins-maven-cd-action/blob/3f1f2fa2739ce0747de9a2545aee5d54917a7abb/settings.xml[bot token for Artifactory].

Note that the deployment skips tests and all other Maven mojos not necessary to produce artifacts, because the GitHub Action first
link:https://github.com/jglick/verify-ci-status-action/blob/7ed33273673b4d6d9d8b0533166dfda072b96807/run.sh[verifies that the Jenkins build passed]
whether the deployment was triggered
link:https://github.com/jenkinsci/log-cli-plugin/blob/b1ee6375b940508709ff64cd8e154ad99eb270ce/.github/workflows/cd.yaml#L2-L4[automatically or manually].
It is the responsibility of the
link:https://github.com/jenkinsci/log-cli-plugin/blob/934670f4a9b84f243f8c0d9bbfc7be30a242e32b/Jenkinsfile#L1[Jenkinsfile]
to ensure that an appropriate set of tests, static analyses, etc. are run on appropriate platforms.

Release Drafter is configured as an Action as well, but always publishing the draft release under the placeholder name `next` rather than a real version.
This runs (fairly) quickly after PRs are merged, and just updates a draft release.
If and when a release is successfully deployed to Artifactory, the deployment script

> link:https://github.com/jglick/jenkins-maven-cd-action/blob/3f1f2fa2739ce0747de9a2545aee5d54917a7abb/run.sh#L5-L8[creates a (lightweight) Git tag and publishes the release].

Reviewer: is the lightweight bit relevant?

Author: Well, somewhat; there is a different GitHub API for creating an annotated tag, which I am not doing.

=== Organization setup

To avoid the need for each component maintainer to manually create a bot account and perform similar setup,
tooling at the organization level automates this aspect.

link:https://github.com/jenkins-infra/repository-permissions-updater[Repository Permissions Updater (RPU) processing]
recognizes a flag in a repo’s YAML entry (`cd.enabled` set to `true`).
If that flag is set, a group whose name is derived from the GitHub repository name is created in Artifactory.
That group is granted the same permissions as regular maintainers, but contains no users.
Instead, RPU periodically generates a temporary Artifactory access token that's a member of that group

> and saves it to the GitHub secrets `ARTIFACTORY_USERNAME` and `ARTIFACTORY_TOKEN` in the corresponding repository.

Reviewer 2: Would it make sense to be more implementation agnostic here and call these `MAVEN_REPO_USERNAME` / `MAVEN_REPO_TOKEN` instead?

Author: Fine with me. If you want to make such a change in RPU and also push fresh secrets with those names to `log-cli-plugin`, I will update everything to match.

to ensure that an appropriate set of tests, static analyses, etc. are run on appropriate platforms.

Release Drafter is configured as an Action as well, but always publishing the draft release under the placeholder name `next` rather than a real version.
This runs (fairly) quickly after PRs are merged, and just updates a draft release.
If and when a release is successfully deployed to Artifactory, the deployment script
link:https://github.com/jenkinsci/log-cli-plugin/blob/d041abc4ff822401430b3bc00bf5da31f1d6e8d9/.github/workflows/release.sh#L6-L10[creates a (lightweight) Git tag and publishes the release].

In the future there may also be a new Action or other tooling wrapping most of `release.yaml`, `release.sh`, and `settings.xml` to avoid duplicating shell scripts across repositories.

=== Organization setup

To avoid the need for each component maintainer to manually create a bot account and perform similar setup,
tooling at the organization level automates this aspect.

link:https://github.com/jenkins-infra/repository-permissions-updater[Repository Permissions Updater (RPU) processing]

> recognizes a flag in a repo’s YAML entry ('cd.enabled` set to `true`)
> and creates an Artifactory bot account with permissions to only the listed paths
> (via a synthetic Artifactory group),
> then generates an access token for that account
> and saves it to the GitHub secrets `ARTIFACTORY_USERNAME` and `ARTIFACTORY_TOKEN` in the corresponding repository.

Reviewer 2: Groups are real (and otherwise unused), accounts are not.

```suggestion
recognizes a flag in a repo’s YAML entry ('cd.enabled` set to `true`).
If that flag is set, a group whose name is derived from the GitHub repository name is created in Artifactory.
That group is granted the same permissions as regular maintainers, but contains no users.
Instead, RPU periodically generates a temporary Artifactory access token that's a member of that group
and saves it to the GitHub secrets `ARTIFACTORY_USERNAME` and `ARTIFACTORY_TOKEN` in the corresponding repository.
```

* Theft of tokens prior to expiry by malicious maintainers is a possibility under either system;
  the window of opportunity would differ, as would the sort of audit trail produced.

=== Release Drafter as a step in the release workflow

> Unfortunately it is not possible to run Release Drafter as a step in the release workflow,

Reviewer 3: https://github.com/jenkinsci/core-changelog-generator has a changelog generator mode which takes the Release Drafter format. If needed, we can use it in the recommended flow inside Jenkins or elsewhere

Reviewer 3: FTR https://github.com/jenkinsci/core-changelog-generator/pull/12 . Some patches are still needed to support all Release Drafter options, the patch was designed specifically for the Jenkins core

Author: Thanks. I think it is safest to stick with using official RD even if it needs to be done as a separate workflow at least for now. Can be revisited as we get real experience though. Will note your idea.

link:https://github.com/jenkins-infra/repository-permissions-updater/pull/1747[repository-permissions-updater #1747]
implements changes to infrastructure.

link:https://github.com/jenkinsci/plugin-pom/pull/375[plugin-pom #375] and link:https://github.com/jenkinsci/pom/pull/147[pom #147]

> add a POM profile to block MRP in CD mode.

Reviewer 2: It is unclear what this is for. Prevent accidental misuse?

Author: Yes, prevents you from running `mvn release:prepare release:perform`, which would produce crazy results. If you opt into `changelist.format`, you agree not to use MRP.
